### PR TITLE
Whedon deposit

### PR DESCRIPTION
**Pull request by arfon**
This PR basically adds a new command for Whedon which adds the ability to carry out the final steps in accepting a paper (compiling the paper, uploading to GitHub, depositing Crossref metadata):

Issuing the following command will do a 'dry run' of the paper acceptance process e.g. https://github.com/openjournals/joss-reviews/issues/1035#issuecomment-431216563:

```
@whedon accept
```

screen shot 2018-10-23 at 10 06 52 am

If everything looks good then pass the `deposit=true` option (e.g. https://github.com/openjournals/joss-reviews/issues/1035#issuecomment-431216779):

```
@whedon accept deposit=true
```

screen shot 2018-10-23 at 10 08 32 am

Note, I've thus far tried out this new functionality on a few JOSS submissions. It would be good to have some of the @openjournals/jose-editors editors take this for a spin sometime soon too.

### CODE REVIEW

**Comment by labarba** [comment]
OMG this functionality rocks my world!

**Comment by kyleniemeyer** [comment]
@labarba I am about to test it out for JOSE! https://github.com/openjournals/jose-reviews/issues/33

**Comment by labarba** [comment]
Please don’t publish a paper without EIC approval.

File: workers.rb [hunk 1]
```diff
@@ -128,63 +119,117 @@ def compile(issue_id, custom_branch=nil)
...
+    ENV['JOURNAL_NAME'] = site_name
+    ENV['JOURNAL_LAUNCH_DATE'] = journal_launch_date
   end
+end
 
-  # GitHub stuff (to be refactored!)
-  def get_master_ref(papers)
-    github_client.refs(papers).select { |r| r[:ref] == "refs/heads/master" }.first.object.sha
-  end
+# This is the Sidekiq worked that processes PDFs. It leverages the Whedon gem to
```
Comment (willingc) [inline workers.rb]: worker

File: workers.rb [hunk 2]
```diff
@@ -128,63 +119,117 @@ def compile(issue_id, custom_branch=nil)
...
+# carry out the majority of its actions. Where possible, we try and capture
+# errors from any of the executed tasks and report them back to the review issue
+class DepositWorker
+  require_relative 'github'
+  require 'open3'
+  require 'sidekiq'
+
+  include Sidekiq::Worker
+
+  # Including this means we can talk to GitHub from the background worker.
```
Comment (willingc) [inline workers.rb]: Perhaps: 
`# Include to communicate from background worker to GitHub`

File: workers.rb [hunk 3]
```diff
@@ -128,63 +119,117 @@ def compile(issue_id, custom_branch=nil)
...
+    set_env(papers_repo, site_host, site_name, journal_alias, journal_issn, journal_launch_date, nwo, crossref_username, crossref_password, whedon_secret)
+
+    # Download the paper
+    stdout, stderr, status = download(issue_id)
+
+    # Whedon often can't find a paper in the repository he's downloaded even
+    # though it's definitely there (e.g. https://github.com/openjournals/joss-reviews/issues/776#issuecomment-397714563)
+    # Not sure if this is because the repository hasn't downloaded yet.
+    # Adding in a sleep statement to see if this helps.
+    sleep(5)
```
Comment (willingc) [inline workers.rb]: Perhaps bump to 10 or 15.

**Comment by labarba** [comment]
@arfon, I think this needs some discussion from the point of view of the editorial process. I don't think that all editors should have the full authority to **publish** a paper. It would be good to discuss how you view the new process, with this power granted to `@whedon`, and how it should be wielded most wisely.

**Comment by kyleniemeyer** [comment]
@labarba @arfon that's a good point. There should probably be a team or something of people (EiC, Assistant EiC) who has this power.

**Comment by arfon** [comment]
> Or at least a sanity check that results in the EiC (and/or Associate EiCs?) giving a final OK ... which given the ease of typing `@whedon accept` might as well mean that only they get to use that command.

:+1: lemme look into this.

**Comment by labarba** [comment]
Will the `@whedon accept` command still require that the **accepted** label be added to the REVIEW issue thread?

**Comment by arfon** [comment]
Actually, Whedon adds that if it isn't already there.

**Comment by arfon** [comment]
So in [this commit](https://github.com/openjournals/whedon-api/pull/41/commits/0c737c603b1ce2293315ae83e9ac64a13c1031be) I've added the functionality to allow us to differentiate between editors and eic-like editors who can actually issue the command `@whedon accept deposit=true`

This change isn't yet deployed but it would basically lock down the final publishing step to a list of defined editors.

What do folks think?

**Comment by kyleniemeyer** [comment]
Yeah, this is a good idea, and I think fits quite nicely with the assistant (associate?) EiC idea

**Comment by kyleniemeyer** [comment]
I do think that there should be a higher tier above the individual editors that does the final accept & publish; that's consistent with how things have been working (with @arfon doing all the work) and also traditional journals.

**Comment by labarba** [comment]
* Accept decisions can be made by any and all editors, but Publish decisions no.
* Ideally, handling editors add the "Accept" label, so publishing editors can identify accepted papers that need attention (Review issue still open).
* Final publication authority is under a high bar of trust, and so limited to EiC and Associate EiC (that's the common term)

**Comment by arfon** [comment]
> - Final publication authority is under a high bar of trust, and so limited to EiC and Associate EiC (that's the common term)

Yep, this makes sense.

**Pull request merged**